repository-s3/src/main/java/io/aiven/elasticsearch/repositories/s3/S3RepositoryStorageIOProvider.java

```diff
@@ -0,0 +1,271 @@
+            } catch (final Exception e) {
+                throw new BlobStoreException("Failed to check if blob [" + blobName + "] exists", e);
+            }
+        }
+
+        @Override
+        public InputStream read(final String blobName) throws IOException {
+            return Permissions.doPrivileged(() -> {
+                final S3ObjectInputStream objectContent;
+                try {
+                    objectContent = client.getObject(bucketName, blobName).getObjectContent();
+                } catch (final AmazonClientException e) {
+                    throw new IOException("Couldn't read blob " + blobName, e);
+                }
+                return cryptoIOProvider.decryptAndDecompress(objectContent);
+            });
+        }
+
+        @Override
+        public void write(final String blobName,
+                          final InputStream inputStream,
+                          final long blobSize,
+                          final boolean failIfAlreadyExists) throws IOException {
+            try {
+                Permissions.doPrivileged(() ->
+                        cryptoIOProvider.compressAndEncrypt(
+                                inputStream,
+                                new S3OutputStream(bucketName, blobName, partSize, client)
+                        )
+                );
+            } catch (final AmazonClientException e) {
+                throw new IOException("Couldn't upload blob with name: " + blobName, e);
+            }
+        }
+
+        @Override
+        public Tuple<Integer, Long> deleteDirectories(final String path) throws IOException {
+            final List<Tuple<String, Long>> listOfFiles;
+            try {
+                //we need only prefix here since it is impossible to distinguish between "files" and "directories"
+                //in AWS SDK e.g. paths like:
+                // - /aaaa/bbb/ccc/file
+                // - /aaa/bbb/file
+                // with request which contains delimiter and without returns diff result, and ES API doesn't use
+                // recursive calls to get full list of files by path/sub-path
+                listOfFiles = Permissions.doPrivileged(() ->
+                        list(new ListObjectsV2Request().withBucketName(bucketName).withPrefix(path)))
+                        .stream()
+                        .flatMap(l -> l.getObjectSummaries().stream())
+                        .map(o -> Tuple.tuple(o.getKey(), o.getSize()))
+                        .collect(Collectors.toList());
+            } catch (final AmazonClientException e) {
+                throw new IOException("Couldn't get list of files for path " + path, e);
+            }
+
+            deleteFiles(listOfFiles.stream()
+                            .map(Tuple::v1)
+                            .collect(Collectors.toList()),
+                    true);
+
+            final var removedBytes = listOfFiles.stream().map(Tuple::v2).reduce(0L, Long::sum);
```

Comment:
I'm not fully aware of how ES uses the deleted blobs and bytes numbers, but I assume this matters somehow. If so, than the result may be inconsistent when deleting non-existing keys. I propose to count a blob and its size to the result only when it's actually deleted.

Reply:
True, will fix

Reply:
DeleteRequest is just response which will contains this data for the end user. Since S3 doesn't use thoiw exception for unknown objects in the bucket. It is quit difficult to understand which files were removed... Any ideas?

Reply:
> DeleteRequest is just response which contains this data for the end user

Well, then it doesn't probably worth the effort